File: B. Singly Linked List/singlyLinkedList.cpp
#include <iostream>
#include <new>
using namespace std;

// Singly linked list class
template <typename T>
class SinglyLinkedList {
  public:
    class Node;
  private:
    int size = 0;
    Node *head = NULL;
    Node *tail = NULL;
  public:
    // Node class
    class Node {
      public:
        T data;
        Node *next;
        // Constructor
        Node(T data = NULL, Node *next = NULL) {
          this->data = data;
          this->next = next;
        }
    };

    // Add a node to the tail of the linked list, O(1)
    void add(T elem) {
      Node *temp = new Node(elem);
      if (size == 0) {
        head = tail = temp;
      } else {
        tail->next = temp;
        tail = tail->next;
      }
      size++;
    }

    // Add an element to the beginning of this linked list, O(1)
    void addFirst(T elem) {
      if (size == 0) {
        head = tail = new Node(elem);
      } else {
        Node *temp = new Node(elem, head);
        head = temp;
      }
      size++;
    }

    // Check the value of the first node if it exists, O(1)
    T peekFirst() {
      try {if (size == 0) throw 101;} catch(...) {cout<<"Empty list";}
      return head.data;
    }

    // Check the value of the last node if it exists, O(1)
    T peekLast() {
      try {if (size == 0) throw 101;} catch(...) {cout<<"Empty list";}
      return tail.data;
    }

    // Remove a node at a particular index, O(n)
    void removeAt(int index) {
      // Make sure the index provided is valid
      try {
        if (index < 0 || index >= size) {
          throw 101;
        }
      } catch(...) {
        cout<<"Invalid index";
      }

      Node *temp = head;
      if (index == 0) {
        head = head->next;
        delete temp;
      }
      else
      {
        // Search from the front of the list
        for (int i = 1; i != index; i++) {
          temp = temp->next;
        }
        Node *toDelete = temp->next;
        temp->next = toDelete->next;
        delete toDelete;
      }
      size--;
    }

    // Remove a particular value in the linked list, O(n)
    bool remove(T value) {
      Node *temp = head;
      if (head->data == value) {
        head = head->next;
        delete temp;
        size--;
        return true;
      }
      else
      {
        // Search for value
        for (; temp->next != NULL; temp = temp->next) {
          if (value == temp->next->data) {
            Node *nextNode = temp->next->next;
            delete temp->next;
            temp->next = nextNode;
            size--;
            return true;
          }
        }
      }
      return false;
    }

    // Find the index of a particular value in the linked list, O(n)
    int indexOf(T value) {
      int index = 0;
      Node *trav = head;
      for (; trav != NULL; trav = trav->next, index++) {
        if (value == trav->data) {
          return index;
        }
      }
      return -1;
    }

    // Check is a value is contained within the linked list
    bool contains(T value) {
      return indexOf(value) != -1;
    }

    // Print values in linked list
    void print() {
      Node *trav = head;
      while (trav != NULL) {
        cout<<trav->data<<" ";
        trav = trav->next;
      }
    }
};

int main()
{
  SinglyLinkedList<int> A;
  A.add(1);
  A.add(2);
  A.add(3);
  A.add(4);
  A.add(5);
  A.add(6);
  A.removeAt(0);
  A.removeAt(1);
  A.remove(5);
  A.print();
  return 0;
}

File: Singly Linked List/singlyLinkedList.py
class SinglyLinkedList:
    # Node class inside linked list
    class Node:
        # Node constructor
        def __init__(self, data = None, next = None):
            self.data = data
            self.next = next

    # Linked list constructor
    def __init__(self):
        self.head = None
        self.size = 0

    # Insert element at last in linked list, O(n)
    def insert(self, data):
        if self.head == None:
            self.head = self.Node(data)
        else:
            temp = self.head
            while temp.next != None:
                temp = temp.next
            temp.next = self.Node(data)
        self.size += 1

    # Remove element at given index in linked list, O(n)
    def removeAt(self, index):
        if index < 0 or index > self.size - 1:
            print('Error')
        elif index == 0:
            temp = self.head
            self.head = temp.next
            del temp
            self.size -= 1
        else:
            temp = self.head
            for i in range(1, index):
                temp = temp.next
            toDelete = temp.next
            temp.next = toDelete.next
            del toDelete
            self.size -= 1

    # Remove last element in linked list, O(n)
    def remove(self):
        self.removeAt(self.size - 1)

    # Returns size of linked list, O(1)
    def getSize(self):
        return self.size

    # Find the index of a particular value in the linked list, O(n)
    def indexOf(self, data):
        index = 0
        temp = self.head
        while index < self.size:
            if temp.data == data:
                return index
            temp = temp.next
            index += 1
        else:
            return -1

    # Check is a value is contained within the linked list
    def contains(self, data):
        return self.indexOf(data) != -1

    # Print linked list
    def show(self):
        temp = self.head
        print('List:', end = ' ')
        while temp != None:
            print(temp.data, end = ' ')
            temp = temp.next

x = SinglyLinkedList()
x.insert(2)
x.insert(3)
x.insert(4)
x.insert(6)
x.insert(7)
x.removeAt(2)
x.remove()
x.show()
